pkg/registry/pod/etcd/etcd.go

```diff
@@ -218,6 +220,38 @@ type LogREST struct {
 	kubeletConn client.ConnectionInfoGetter
 }
 
+// Max length read from the body of a log location which returns error status
+const (
+	maxReadLength = 1000
+)
+
+// A log location response checker to transform the error.
+type LogLocationResponseChecker struct {
+	podName string
+}
+
+func (checker LogLocationResponseChecker) Check(resp *http.Response) error {
+	if resp.StatusCode < http.StatusOK || resp.StatusCode > http.StatusPartialContent {
+		defer resp.Body.Close()
+		bodyBytes, err := ioutil.ReadAll(io.LimitReader(resp.Body, maxReadLength))
+		if err != nil {
+			return errors.NewInternalError(err)
+		}
+		bodyText := string(bodyBytes)
+
+		switch {
+		case resp.StatusCode == http.StatusInternalServerError:
+			return errors.NewInternalError(fmt.Errorf("%s", bodyText))
```

Comment:
I do agree fitting our error pattern is good, though.

On Wed, Sep 23, 2015 at 3:31 PM, Daniel Smith <dbsmith@google.com> wrote:

> I'm just imagining getting bug reports for this and having a hard time
> figuring out if someone has a busted apiserver or rouge node...
>
> On Wed, Sep 23, 2015 at 3:28 PM, Clayton Coleman <notifications@github.com
> > wrote:
>
>> In pkg/registry/pod/etcd/etcd.go
>> <https://github.com/kubernetes/kubernetes/pull/13722#discussion_r40266227>
>> :
>>
>> > +type LogLocationResponseChecker struct {
>> > +	podName string
>> > +}
>> > +
>> > +func (checker LogLocationResponseChecker) Check(resp *http.Response) error {
>> > +	if resp.StatusCode < http.StatusOK || resp.StatusCode > http.StatusPartialContent {
>> > +		defer resp.Body.Close()
>> > +		bodyBytes, err := ioutil.ReadAll(io.LimitReader(resp.Body, maxReadLength))
>> > +		if err != nil {
>> > +			return errors.NewInternalError(err)
>> > +		}
>> > +		bodyText := string(bodyBytes)
>> > +
>> > +		switch {
>> > +		case resp.StatusCode == http.StatusInternalServerError:
>> > +			return errors.NewInternalError(fmt.Errorf("%s", bodyText))
>>
>> Hrm - although practically as a client you want to be agnostic to the
>> backend. Returning gateway makes the client have to deal with more things -
>> they may not actually care which server is returning the error. In fact,
>> pod logs returning Kubelet errors is probably not the API we want - we want
>> it to fit the Kube API pattern.
>> … <#14ffc55a835101bc_14ffc52f0a730019_>
>> On Wed, Sep 23, 2015 at 6:27 PM, Daniel Smith <notifications@github.com>
>> wrote: In pkg/registry/pod/etcd/etcd.go <
>> https://github.com/kubernetes/kubernetes/pull/13722#discussion_r40266046>
>> : > +type LogLocationResponseChecker struct { > + podName string > +} > + >
>> +func (checker LogLocationResponseChecker) Check(resp *http.Response) error
>> { > + if resp.StatusCode < http.StatusOK || resp.StatusCode >
>> http.StatusPartialContent { > + defer resp.Body.Close() > + bodyBytes, err
>> := ioutil.ReadAll(io.LimitReader(resp.Body, maxReadLength)) > + if err !=
>> nil { > + return errors.NewInternalError(err) > + } > + bodyText :=
>> string(bodyBytes) > + > + switch { > + case resp.StatusCode ==
>> http.StatusInternalServerError: > + return
>> errors.NewInternalError(fmt.Errorf("%s", bodyText)) I'm worried this is
>> going to be very confusing, now that I think about it. I think that, no
>> matter what error happened downstream, we need to send a "bad gateway"
>> error upstream to the original client. If we do this, the client can't tell
>> whether *we* had the 500 or the downstream server had the 500. — Reply to
>> this email directly or view it on GitHub <
>> https://github.com/kubernetes/kubernetes/pull/13722/files#r40266046>.
>>
>> —
>> Reply to this email directly or view it on GitHub
>> <https://github.com/kubernetes/kubernetes/pull/13722/files#r40266227>.
>>
>
>